Add a new jobs endpoint to serve status information for async requests

julianvilas opened the pull request:
This endpoint will be used to retrieve information about the status of async requests being processed in the background.

Modifications introduced:

* A new database table to store jobs, with these fields:
```| id | team_id | operation | status | result | created_at | updated_at |```
* A new global endpoint to retrieve a job by its ID:
```GET /api/v1/jobs/{job_id}/```

ka3de commented on pkg/api/job.go:
```diff
@@ -0,0 +1,39 @@
+// Job contains the status information of an asynchronous operation.
+//
+// In case of non-global operations it also contains the team ID associated to
+// the operation.
```
This comment made me think that this "non-global" operation is not going to be possible according to the table definition we made because we are using a foreign key on `teams(id)`. We should probably discuss and rethink about this, if this "global-operations" are gonna be a valid use case and how to handle that.

ka3de commented on pkg/api/service/jobs.go:
```diff
@@ -0,0 +1,19 @@
+func (s vulcanitoService) FindJob(ctx context.Context, id string) (*api.Job, error) {
+	if id == "" {
+		return nil, errors.Validation(`ID is empty`)
+	}
```
No big deal, and we can keep it here, but do you think this validation would make more sense at the endpoint level?

julianvilas replied:
Yes, absolutely. I thought it was done but it wasn't.

julianvilas replied:
Done

ka3de reviewed (changes requested):
Hey @julianvilas good job! :+1: 
Overall everything looks very good but I think there are some minor things to take a look at and a discussion pending about the support for "global jobs" (see my comment on `pk/api/job.go`).

Additionally I believe we are missing the endpoint "entry" in he metrics middleware (`pkg/api/middleware/metrics.go`), for which I believe we'll have to add a new type of entity for `job`.

julianvilas commented on cmd/vulcan-api-cli/design/jobs.go:
```diff
@@ -0,0 +1,42 @@
+	})
+})
+
+var _ = Resource("job", func() {
```
Totally agree. Will change it.

julianvilas replied:
Done!

julianvilas commented on pkg/api/store/job.go:
```diff
@@ -0,0 +1,35 @@
+	}
+
+	if res.RowsAffected == 0 {
+		return nil, db.logError(errors.NotFound("Job does not exists"))
```
Done!

julianvilas commented on cmd/vulcan-api-cli/design/jobs.go:
```diff
@@ -0,0 +1,42 @@
+	DefaultMedia(UserMedia)
+
+	Action("show", func() {
+		Description("Describe job")
```
Sure! I'm on it.

julianvilas replied:
Done!

ka3de commented on pkg/api/job.go:
```diff
@@ -15,25 +17,54 @@ import (
+	// - DONE
+	Status string `validate:"required"`
+	// Result is a marshaled JSON representation of a JobResult.
+	Result string `gorm:"Column:result"`
```
Should we use the `JobResult` struct here to represent the `Result` field?
I have some doubts about how Gorm will treat this when inserting into DB when using a struct, so we can check and discuss.

julianvilas replied:
Let's discuss it in the meeting we have and choose whatever fits better :)

julianvilas replied:
We went for the JSONB approach finally.

ka3de commented on pkg/api/job.go:
```diff
@@ -15,25 +17,54 @@ import (
+	case "RUN":
+	case "DONE":
+	default:
+		return errors.New("valid states are PENDING, RUN or DONE")
```
In reference to the comment above, should we use `RUN` or `RUNNING` as status?
Also, super silly but, I guess it's correct to use `state` and `status` interchangeably in English, but should we s/states/status here for coherence?

julianvilas replied:
Done!

ka3de commented on db/sql/V1.51__create_jobs_table.sql:
```diff
@@ -3,6 +3,7 @@ CREATE TABLE jobs (
 	team_id UUID,
     operation TEXT NOT NULL,
     status TEXT NOT NULL,
+    result TEXT,
```
Should we use `jsonb` type here? That will allow us to take advantage of PSQL capabilities to interact with JSON data. Also I guess Gorm will provide better support for our implementation in that way too.

julianvilas replied:
Let's discuss it in the meeting we have and choose whatever fits better :)

julianvilas replied:
We went for the JSONB approach finally.

ka3de reviewed (commented):
I @julianvilas great job! :+1: 
Thanks for addressing the previous comments, I think we have a better idea now of what we want to achieve.
I made some comments, mostly simple things except for the comments about using JSONB type for Job result field. We can discuss this and take a look together to see what is the best approach.

julianvilas commented on cmd/vulcan-api-cli/design/jobs.go:
```diff
@@ -0,0 +1,65 @@
+
+	Action("show", func() {
+		Description(`Describes job status and results. The possible values for the status are:
+	- 'PENDING': The job has been noted an is pending to be processed
```
Done!

julianvilas commented on cmd/vulcan-api-cli/design/jobs.go:
```diff
@@ -0,0 +1,65 @@
+		Description(`Describes job status and results. The possible values for the status are:
+	- 'PENDING': The job has been noted an is pending to be processed
+	- 'RUNNING': The job is on execution
+	- 'DONE': The job has finished, either successfully os unsuccesfully. Result.error needs to be processed to determine it
```
Done!

julianvilas commented on pkg/api/job.go:
```diff
@@ -15,25 +17,54 @@ import (
 	TeamID    string `gorm:"Column:team_id"`
-	Operation string
-	Status    string `validate:"required"`
+	Operation string `validate:required`
```
Done!

julianvilas commented on pkg/api/job.go:
```diff
@@ -15,25 +17,54 @@ import (
+func (j Job) Validate() error {
+	switch j.Status {
+	case "PENDING":
+	case "RUN":
```
Done!

julianvilas commented on pkg/api/job.go:
```diff
@@ -15,25 +17,54 @@ import (
+	switch j.Status {
+	case "PENDING":
+	case "RUN":
+	case "DONE":
```
Let's discuss it together with the JSONB approach in the meeting we have and choose whatever fits better :)

julianvilas replied:
We went for the constants option. In the future we might explore using enums both in Go code and in the database table.

julianvilas commented:
Thanks for your review and comments @ka3de :) 

I've addressed all your comments excepts those we plan to discuss in a meeting: the JSONB Results and the JobStatus topics. Thanks again!

ka3de reviewed (approved):
LGTM.
